cli/upload.py

```diff
@@ -83,65 +84,154 @@ def _open_file_mapping(extra_metadata: dict) -> Dict[str, BinaryIO]:
             f.close()
 
 
+_IGNORED_WARN_LINES = set(
+    map(
+        str.strip,
+        """==> NOTE: You are uploading one or more large file(s), which would run
+significantly faster if you enable parallel composite uploads. This
+feature can be enabled by editing the
+"parallel_composite_upload_threshold" value in your .boto
+configuration file. However, note that if you do this large files will
+be uploaded as `composite objects
+<https://cloud.google.com/storage/docs/composite-objects>`_,which
+means that any user who downloads such objects will need to have a
+compiled crcmod installed (see "gsutil help crcmod"). This is because
+without a compiled crcmod, computing checksums on composite objects is
+so slow that gsutil disables downloads of composite objects.""".split(
+            "\n"
+        ),
+    )
+)
+
+
+def _start_procs(upload_info: api.UploadInfo, xlsx: str) -> list:
+    """
+    Starts multiple "gsutil cp" subprocesses.
+    Returns a list of subprocess.Popen objects
+    """
+    procs = []
+
+    for src, dst in _compose_file_mapping(upload_info, xlsx):
+
+        # Construct the upload command
+        gsutil_args = ["gsutil", "-m", "cp", src, dst]
+
+        try:
+            # Run the upload command
+            procs.append(
+                subprocess.Popen(
+                    gsutil_args,
+                    universal_newlines=True,
+                    bufsize=1,  # line buffered so we can read output line by line
+                    stdout=subprocess.DEVNULL,
+                    stderr=subprocess.PIPE,
+                )
+            )
+
+        except OSError as e:
+
+            # stopping already created processes
+            for p in procs:
+                p.kill()
+
+            _handle_upload_exc(e)
+    return procs
+
+
+def _wait_for_upload(procs: list) -> bool:
+    """
+    Waits for all subprocesses and prints their stderr streams.
+    Returns bool - if an error has occurred during any if uploads
+    """
+
+    finished = set()
+    errored = False
+    while len(finished) != len(procs) and not errored:
```

Comment:
Is there a way to wait for subprocesses to finish without spinning in this way? If not, should this loop at least sleep briefly between iterations?

Reply:
the idea is that `.stderr.readline()` (few lines down) will block until gsutil prints next "progress" line, which is couple of times per second, so this won't be a hot loop, I'd say.

I confirmed in a manual measurement that the loop waits in the `readline` call 99% of the time. 

What do you think?

Reply:
Sounds fine to me!